middleware/middleware.go

```diff
@@ -0,0 +1,75 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/services/stats"
+)
+
+func MaxConcurrentRequestsMiddleware(maxConcurrentRequests int) func(http.Handler) http.Handler {
+
+	var requests = make(chan struct{}, maxConcurrentRequests)
+	return func(next http.Handler) http.Handler {
+		//this is to make sure that we don't have more than `maxClient` in-memory at any point of time. As, having more http client than `maxClient`
+		// may lead to gateway OOM kill.
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if maxConcurrentRequests != 0 {
+				select {
+				case requests <- struct{}{}:
+					defer func() {
+						<-requests
+					}()
+				default:
+					http.Error(w, http.StatusText(http.StatusServiceUnavailable), http.StatusServiceUnavailable)
+					return
+				}
+			}
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func StatMiddleware(ctx context.Context) func(http.Handler) http.Handler {
+	var concurrentRequests int32
+	activeClientCount := stats.NewStat("gateway.concurrent_requests_count", stats.GaugeType)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(10 * time.Second):
+				activeClientCount.Gauge(concurrentRequests)
+			}
+		}
+	}()
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			latencyStat := stats.DefaultStats.NewSampledTaggedStat("gateway.response_time", stats.TimerType, map[string]string{"reqType": r.URL.Path})
+			latencyStat.Start()
+			defer latencyStat.End()
+
+			atomic.AddInt32(&concurrentRequests, 1)
+			defer atomic.AddInt32(&concurrentRequests, -1)
+
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+//Currently sets the content-type only for eventSchemas, health responses.
+//Note : responses via http.Error aren't affected. They default to text/plain
+func HeaderMiddleware() func(http.Handler) http.Handler {
```

Comment:
[minor] 
1. Include the function name in the comment.
2. Suggest a more appropriate title. `ContentType`

```suggestion
// ContentType Currently sets the content-type only for eventSchemas, health responses.
// Note : responses via http.Error aren't affected. They default to text/plain
func ContentType() func(http.Handler) http.Handler {
```

Reply:
done